src/util/system.cpp

```diff
@@ -1111,6 +1102,47 @@ void AllocateFileRange(FILE *file, unsigned int offset, unsigned int length) {
 #endif
 }
 
+FILE* AdviseSequential(FILE *file) {
+#if _POSIX_C_SOURCE >= 200112L
+    // Since this whole thing is advisory anyway, we can ignore any errors
+    // encountered up to and including the posix_fadvise call. However, we must
+    // rewind the file to the appropriate position if we've changed the seek
+    // offset.
+    if (file == nullptr)
+        return nullptr;
+    int fd = fileno(file);
+    if (fd == -1)
+        return file;
+    off_t start = lseek(fd, 0, SEEK_CUR);
+    if (start == -1)
+        return file;
+    off_t end = lseek(fd, 0, SEEK_END);
+    if (end != -1) {
+        posix_fadvise(fd, start, end - start, POSIX_FADV_WILLNEED);
+        posix_fadvise(fd, start, end - start, POSIX_FADV_SEQUENTIAL);
+    }
+    lseek(fd, start, SEEK_SET);
+#endif
+    return file;
+}
+
+int CloseAndUncache(FILE *file) {
+#if _POSIX_C_SOURCE >= 200112L
+    // Ignore any errors up to and including the posix_fadvise call since it's
+    // advisory.
+    if (file != nullptr) {
+        const int fd = fileno(file);
+        if (fd != -1) {
+            const off_t end = lseek(fd, 0, SEEK_END);
```

Comment:
nit: I like the `const`, but it is missing from variables in `AdviseSequential()`, maybe add it there too?